client/components/tinymce/plugins/mentions/mentions.jsx

```diff
@@ -0,0 +1,164 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+import ReactDomServer from 'react-dom/server';
+import { connect } from 'react-redux';
+import { includes } from 'lodash';
+import tinymce from 'tinymce/tinymce';
+
+/**
+ * Internal dependencies
+ */
+import SuggestionList from './suggestion-list';
+import EditorMention from './editor-mention';
+import QueryUsersSuggestions from 'components/data/query-users-suggestions';
+import { getSelectedSiteId } from 'state/ui/selectors';
+import { getUserSuggestions } from 'state/users/suggestions/selectors';
+
+/**
+ * Module variables
+ */
+const VK = tinymce.util.VK;
+
+const getMatchingSuggestions = function( suggestions, query ) {
+	const matcher = new RegExp( '^' + query + '| ' + query, 'ig' ); // Start of string or preceded by a space.
+	let matchingSuggestions = suggestions;
+
+	if ( ( query !== null ) && ( query.length > 0 ) ) {
+		matchingSuggestions = [];
+
+		for ( let i = 0, len = suggestions.length; i < len; i++ ) {
```

Comment:
This would be a good candidate for an [`Array#filter`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter).

```js
matchingSuggestions = suggestions.filter( ( suggestion ) => {
	const name = suggestion.name || suggestion.user_login + ' ' + suggestion.display_name;
	matcher.test( name );
} );
```

Note also in this example I've removed `toLowerCase`. Since your regular expression specifies [`i` flag](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp), you shouldn't need this.

One more optimization I could imagine would be to stop filtering after 10 results have been found, since the `Array#slice` below will result in only the first 10 being used anyways. It's not obvious how to achieve this, outside of using a `for` loop with `break;` or perhaps [Lodash's `_.forEach`](https://lodash.com/docs/4.16.6#forEach) special treatment of `return false;`. Probably not worth it at the loss of readability.